examples/sir_hmc.py

```diff
@@ -0,0 +1,434 @@
+        S = S_next
+        I = I_next
+        pyro.sample("obs_{}".format(t),
+                    dist.ExtendedBinomial(S2I.clamp(min=0), rho),
+                    obs=datum)
+
+
+# Now all latent variables in the continuous_model are either continuous or
+# enumerated, so we can use HMC. However we need to take special care with
+# constraints because the above Markov reparameterization covers regions of
+# hypothesis space that are infeasible (i.e. whose log_prob is -infinity). We
+# thus heuristically initialize to a feasible point.
+
+def heuristic_init(args, data):
+    """Heuristically initialize to a feasible point."""
+    # Start with a single infection.
+    S0 = args.population - 1
+    if args.heuristic == 1:
+        # Assume >= 50% response rate.
+        S2I = data * min(2., S0 / data.sum())
+        S_aux = (S0 - S2I.cumsum(-1)).clamp(min=0.5)
+        # Account for the single initial infection.
+        S2I[0] += 1
+        # Assume infection lasts less than a month.
+        recovery = (1 - args.recovery_rate / 2) ** torch.arange(30.)
+        I_aux = convolve(S2I, recovery)[:len(data)].clamp(min=0.5)
+    elif args.heuristic == 2:
+        # Assume 100% response rate, 0% recovery rate.
+        S_aux = S0 - data.cumsum(-1)
+        I_aux = 1 + data.cumsum(-1)
+    # Also initialize DCT transformed coordinates.
+    t = ComposeTransform([biject_to(constraints.interval(-0.5, args.population + 0.5)).inv,
+                          DiscreteCosineTransform(dim=-1)])
+
+    return {
+        "prob_s": torch.tensor(0.5),
+        "prob_i": torch.tensor(0.5),
+        "rho": torch.tensor(0.5),
+        "S_aux": S_aux,
+        "I_aux": I_aux,
+        "S_aux_dct": t(S_aux),
+        "I_aux_dct": t(I_aux),
+    }
+
+
+# One trick to improve inference geometry is to reparameterize the S_aux,I_aux
+# variables via DiscreteCosineReparam. This allows HMC's diagonal mass matrix
+# adaptation to learn different step sizes for high- and low-frequency
+# directions. We can apply that outside of the model, during inference.
+
+def infer_hmc_cont(model, args, data):
+    if args.dct:
+        model = poutine.reparam(model, {"S_aux": DiscreteCosineReparam(),
+                                        "I_aux": DiscreteCosineReparam()})
+    init_values = heuristic_init(args, data)
+    return _infer_hmc(args, data, model, init_values=init_values)
+
+
+# The next step is to vectorize. We can repurpose DiscreteHMM's implementation
+# here, but we'll need to manually represent a Markov neighborhood of multiple
+# Bernoullis as single joint Categorical with 2 x 2 = 4 states.
+
+def quantize_enumerate(x_real, min, max):
+    """Quantize, then manually enumerate."""
+    lb = x_real.detach().floor()
+    # This cubic spline guarantees gradients are continuous.
+    t = x_real - lb
+    prob = t * t * (3 - 2 * t)
+    x = torch.stack([lb, lb + 1], dim=-1).clamp(min=min, max=max)
+    logits = safe_log(torch.stack([1 - prob, prob], dim=-1))
```

Comment:
In Bernoulli distribution, prob is clamped to (eps, 1-eps). So you might want to do so here.

Reply:
Nevermind, this is not important.